### PR TITLE
Add `reason` tag to `kubernetes_state.job.failed`

### PR DESCRIPTION
**Pull request by keisku**
### What does this PR do?

Add `reason:backofflimitexceeded,deadlineexceeded` to the `kubernetes_state.job.failed`.

### Motivation

Currently, we cannot monitor why a job failed.

### Additional Notes

- [Job termination and cleanup](https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-termination-and-cleanup)
- BackoffLimitExceeded and DeadlineExceeded are GA. Other reasons are still beta or alpha.
  - https://github.com/kubernetes/kubernetes/blob/f6fde3ba7fb0755eb0d982b6ed2617258f1f4f8e/staging/src/k8s.io/api/batch/v1/types.go#L638-L642
- `kube_job_status_failed` in KSM([code](https://github.com/kubernetes/kube-state-metrics/blob/9e6e70d328c5d750a0df84cc044c38124b271538/internal/store/job.go#L203-L250)) is the only metric that contains `reason` in labels. `kube_job_failed` doesn't have `reason` label, see https://github.com/kubernetes/kube-state-metrics/issues/2382
- `kube_pod_container_status_terminated_reason` in KSM ([code](https://github.com/kubernetes/kube-state-metrics/blob/9e6e70d328c5d750a0df84cc044c38124b271538/internal/store/pod.go#L503-L527)) does not have `reason:backofflimitexceeded,deadlineexceeded` even it can have `reason:ContainerStatusUnknown`, etc.

### Describe how to test/QA your changes

Collect KSM metrics from this job.

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: exit-fail
spec:
  backoffLimit: 2
  template:
    spec:
      containers:
      - name: busybox
        image: busybox
        command: ["exit", "1"]
      restartPolicy: OnFailure
```

### CODE REVIEW

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go [hunk 1]
```diff
@@ -426,7 +428,7 @@ func validateJob(val float64, tags []string) ([]string, bool) {
 	kubeCronjob := ""
 	for _, tag := range tags {
 		split := strings.Split(tag, ":")
-		if len(split) == 2 && split[0] == "kube_job" || split[0] == "job" || split[0] == "job_name" {
+		if len(split) == 2 && split[0] == "kube_job" || split[0] == "job" || split[0] == "job_name" || split[0] == "reason" {
 			// Trim the timestamp suffix to avoid high cardinality
 			if name, trimmed := trimJobTag(split[1]); trimmed {
```
Comment (L3n41c) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go:433]: These 3 lines seem to expect a colon-separated tag whose key is `reason` to have a value which would be a job name.

**Review by hestonhoffman — APPROVED** [review]
Docs 👍

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go [hunk 2]
```diff
@@ -482,10 +498,6 @@ func jobStatusFailedTransformer(s sender.Sender, name string, metric ksmstore.DD
 		return
 	}
 
-	if reasonTagIndex != -1 {
-		tags = append(tags[:reasonTagIndex], tags[reasonTagIndex+1:]...)
-	}
-
```
Comment (keisku) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go:488]: This logic is now in `validateJob()`.

**Review by hestonhoffman — APPROVED** [review]

**Review by sblumenthal — APPROVED** [review]

**Pull request merged**